File: tests/06-advanced-tests.ts
import * as anchor from "@anchor-lang/core";
import { Program } from "@anchor-lang/core";
import { Keypair, PublicKey } from "@solana/web3.js";
import {
  createAssociatedTokenAccount,
  createMint,
  getAccount,
  getMint,
  mintTo,
  TOKEN_2022_PROGRAM_ID,
  TOKEN_PROGRAM_ID,
} from "@solana/spl-token";
import assert from "assert";

import { R3SessionKeys } from "../target/types/r3_session_keys";
import * as sessionKeys from "./utils/pda";
import {
  ensureProgramInitialized,
  executeWithSession,
  timestampFromFuture,
  loadProgramFromFile,
  sendExpectError,
} from "./utils/helpers";
import { DEPOSIT_AMOUNT, INITIAL_TOKEN_BALANCE, TOKEN_DECIMALS } from "./utils/constants";
import { MockProgram, mockProgramIdl } from "./fixtures/mock-program";
import * as mock from "./fixtures/mock-program";

interface TokenPoolFixture {
  depositMint: PublicKey;
  lpMint: PublicKey;
  vault: PublicKey;
  userDepositAccount: PublicKey;
  userLpAccount: PublicKey;
}

describe("06 - Advanced tests (token pool)", () => {
  const provider = anchor.AnchorProvider.env();
  anchor.setProvider(provider);
  const baseWallet = provider.wallet as anchor.Wallet;
  const connection = provider.connection;

  const program = anchor.workspace.r3SessionKeys as Program<R3SessionKeys>;
  let mockProgram: Program<MockProgram>;

  before(async () => {
    await ensureProgramInitialized(program, baseWallet);

    await loadProgramFromFile(connection, mock.MOCK_PROGRAM_ID, mock.MOCK_PROGRAM_SO_PATH);
    mockProgram = new Program(mockProgramIdl as MockProgram, provider) as Program<MockProgram>;
  });

  async function initializeMockTokenPool(tokenOwner: PublicKey): Promise<TokenPoolFixture> {
    const depositMint = await createMint(
      connection,
      baseWallet.payer,
      baseWallet.publicKey,
      null,
      TOKEN_DECIMALS,
      undefined,
      undefined,
      TOKEN_PROGRAM_ID
    );

    const userDepositAccount = await createAssociatedTokenAccount(
      connection,
      baseWallet.payer,
      depositMint,
      tokenOwner,
      undefined,
      TOKEN_PROGRAM_ID
    );

    await mintTo(
      connection,
      baseWallet.payer,
      depositMint,
      userDepositAccount,
      baseWallet.publicKey,
      INITIAL_TOKEN_BALANCE,
      [],
      undefined,
      TOKEN_PROGRAM_ID
    );

    const pool = mock.derivePoolPda(mockProgram.programId, depositMint);
    const lpMint = mock.deriveLpMintPda(mockProgram.programId, pool);
    const vault = mock.deriveVaultAddress(pool, depositMint);

    await mockProgram.methods
      .initializePool()
      .accounts({
        payer: baseWallet.publicKey,
        depositMint,
        depositTokenProgram: TOKEN_PROGRAM_ID,
      })
      .signers([baseWallet.payer])
      .rpc()
      .catch((e) => {
        console.log("Error: ", e);
        throw e;
      });

    // LP mint is created by the mock program as a Token-2022 mint
    const userLpAccount = await createAssociatedTokenAccount(
      connection,
      baseWallet.payer,
      lpMint,
      tokenOwner,
      undefined,
      TOKEN_2022_PROGRAM_ID
    );

    return { depositMint, lpMint, vault, userDepositAccount, userLpAccount };
  }

  async function createApprovedSession(
    allowedInstructionsDiscriminators: Buffer,
    discriminatorLen: number
  ): Promise<{ sessionKey: Keypair; userSmartWallet: PublicKey }> {
    const smartWalletOwner = Keypair.generate();
    const sessionKey = Keypair.generate();

    const userSmartWallet = sessionKeys.deriveUserSmartWalletPda(
      program.programId,
      smartWalletOwner.publicKey
    );
    await program.methods
      .createSmartWallet(smartWalletOwner.publicKey)
      .accounts({
        admin: baseWallet.publicKey,
      })
      .signers([baseWallet.payer])
      .rpc()
      .catch((e) => {
        console.log("Error: ", e);
        throw e;
      });

    await program.methods
      .createSession(
        sessionKey.publicKey,
        mockProgram.programId,
        await timestampFromFuture(connection),
        allowedInstructionsDiscriminators,
        discriminatorLen
      )
      .accounts({
        sessionExecutor: baseWallet.publicKey,
        userSmartWallet,
      })
      .signers([baseWallet.payer])
      .rpc()
      .catch((e) => {
        console.log("Error: ", e);
        throw e;
      });

    await program.methods
      .approveSession(sessionKey.publicKey)
      .accounts({
        smartWalletOwner: smartWalletOwner.publicKey,
      })
      .signers([baseWallet.payer, smartWalletOwner])
      .rpc()
      .catch((e) => {
        console.log("Error: ", e);
        throw e;
      });

    return { sessionKey, userSmartWallet };
  }

  function depositInstruction(user: PublicKey, tokenPool: TokenPoolFixture, amount: number) {
    return mockProgram.methods
      .deposit(new anchor.BN(amount))
      // pool seeds reference its own `deposit_mint` field so it cannot be derived by the resolver,
      // the remaining pool accounts are resolved from it through the IDL relations
      .accountsPartial({
        user,
        pool: mock.derivePoolPda(mockProgram.programId, tokenPool.depositMint),
        userDepositAccount: tokenPool.userDepositAccount,
        userLpAccount: tokenPool.userLpAccount,
        depositTokenProgram: TOKEN_PROGRAM_ID,
      })
      .instruction();
  }

  function withdrawInstruction(user: PublicKey, tokenPool: TokenPoolFixture, amount: number) {
    return mockProgram.methods
      .withdraw(new anchor.BN(amount))
      // pool seeds reference its own `deposit_mint` field so it cannot be derived by the resolver,
      // the remaining pool accounts are resolved from it through the IDL relations
      .accountsPartial({
        user,
        pool: mock.derivePoolPda(mockProgram.programId, tokenPool.depositMint),
        userDepositAccount: tokenPool.userDepositAccount,
        userLpAccount: tokenPool.userLpAccount,
        depositTokenProgram: TOKEN_PROGRAM_ID,
      })
      .instruction();
  }

  async function tokenBalances(tokenPool: TokenPoolFixture) {
    const userDepositAccount = await getAccount(
      connection,
      tokenPool.userDepositAccount,
      undefined,
      TOKEN_PROGRAM_ID
    );
    const vault = await getAccount(connection, tokenPool.vault, undefined, TOKEN_PROGRAM_ID);
    const userLpAccount = await getAccount(
      connection,
      tokenPool.userLpAccount,
      undefined,
      TOKEN_2022_PROGRAM_ID
    );
    const lpMint = await getMint(connection, tokenPool.lpMint, undefined, TOKEN_2022_PROGRAM_ID);

    return {
      userDepositAccount: userDepositAccount.amount,
      vault: vault.amount,
      userLpAccount: userLpAccount.amount,
      lpMintSupply: lpMint.supply,
    };
  }

  it("Execute mock deposit with session executor tokens", async () => {
    const { sessionKey, userSmartWallet } = await createApprovedSession(
      mock.DEPOSIT_DISCRIMINATOR,
      mock.DEPOSIT_DISCRIMINATOR.length
    );

    const sessionExecutor = baseWallet.publicKey;
    const tokenPool = await initializeMockTokenPool(sessionExecutor);
    const depositAccountBefore = await getAccount(
      connection,
      tokenPool.userDepositAccount,
      undefined,
      TOKEN_PROGRAM_ID
    );
    assert.equal(depositAccountBefore.owner.toBase58(), sessionExecutor.toBase58());
    assert.equal(depositAccountBefore.amount, BigInt(INITIAL_TOKEN_BALANCE));

    const depositIx = await depositInstruction(sessionExecutor, tokenPool, DEPOSIT_AMOUNT);
    assert.ok(depositIx.data.subarray(0, 8).equals(mock.DEPOSIT_DISCRIMINATOR));
    // The executor supplies its own tokens. The smart-wallet account is forwarded
    // read-only so the session program can still validate the session's wallet.
    depositIx.keys.push({ pubkey: userSmartWallet, isSigner: false, isWritable: false });

    await executeWithSession(program, {
      sessionExecutor,
      sessionKey: sessionKey.publicKey,
      userSmartWallet,
      targetInstruction: depositIx,
    })
      .signers([baseWallet.payer, sessionKey])
      .rpc()
      .catch((e) => {
        console.log("Error: ", e);
        throw e;
      });

    const balances = await tokenBalances(tokenPool);
    assert.equal(balances.userDepositAccount, BigInt(INITIAL_TOKEN_BALANCE - DEPOSIT_AMOUNT));
    assert.equal(balances.vault, BigInt(DEPOSIT_AMOUNT));
    assert.equal(balances.userLpAccount, BigInt(DEPOSIT_AMOUNT));
    assert.equal(balances.lpMintSupply, BigInt(DEPOSIT_AMOUNT));
  });

  it("Execute mock withdraw not allowed by deposit only session", async () => {
    const { sessionKey, userSmartWallet } = await createApprovedSession(
      mock.DEPOSIT_DISCRIMINATOR,
      mock.DEPOSIT_DISCRIMINATOR.length
    );

    const sessionExecutor = baseWallet.publicKey;
    const tokenPool = await initializeMockTokenPool(sessionExecutor);

    const depositIx = await depositInstruction(sessionExecutor, tokenPool, DEPOSIT_AMOUNT);
    depositIx.keys.push({ pubkey: userSmartWallet, isSigner: false, isWritable: false });
    await executeWithSession(program, {
      sessionExecutor,
      sessionKey: sessionKey.publicKey,
      userSmartWallet,
      targetInstruction: depositIx,
    })
      .signers([baseWallet.payer, sessionKey])
      .rpc()
      .catch((e) => {
        console.log("Error: ", e);
        throw e;
      });

    const balancesBefore = await tokenBalances(tokenPool);

    const withdrawIx = await withdrawInstruction(sessionExecutor, tokenPool, DEPOSIT_AMOUNT);
    assert.ok(withdrawIx.data.subarray(0, 8).equals(mock.WITHDRAW_DISCRIMINATOR));
    assert.ok(!withdrawIx.data.subarray(0, 8).equals(mock.DEPOSIT_DISCRIMINATOR));
    withdrawIx.keys.push({ pubkey: userSmartWallet, isSigner: false, isWritable: false });

    const error = await sendExpectError(
      executeWithSession(program, {
        sessionExecutor,
        sessionKey: sessionKey.publicKey,
        userSmartWallet,
        targetInstruction: withdrawIx,
      })
        .signers([baseWallet.payer, sessionKey])
        .rpc()
    );

    assert.ok(error.includes("NotAllowedInstructionDiscriminator"), error);
    const balancesAfter = await tokenBalances(tokenPool);
    assert.deepEqual(balancesAfter, balancesBefore);
  });
});

File: programs/r3-session-keys/tests/06-advanced_tests.rs
mod common;

use {
    anchor_lang::{
        prelude::Pubkey,
        solana_program::{program_pack::Pack, system_instruction},
        AccountDeserialize,
    },
    anchor_spl::{
        associated_token,
        token::{self, spl_token},
        token_2022,
        token_interface::{Mint, TokenAccount},
    },
    common::{client, mock_client, send_tx_expect_error, send_tx_expect_ok, Env},
    solana_keypair::Keypair,
    solana_signer::Signer,
    std::time::{SystemTime, UNIX_EPOCH},
    associated_token::spl_associated_token_account::instruction::create_associated_token_account,
};

const TOKEN_DECIMALS: u8 = 6;
const INITIAL_TOKEN_BALANCE: u64 = 1_000_000;
const DEPOSIT_AMOUNT: u64 = 250_000;

struct TokenPoolFixture {
    deposit_mint: Pubkey,
    lp_mint: Pubkey,
    vault: Pubkey,
    user_deposit_account: Pubkey,
    user_lp_account: Pubkey,
}

fn load_mock_program(env: &mut Env) {
    let bytes = include_bytes!(concat!(
        env!("CARGO_MANIFEST_DIR"),
        "/../../tests/fixtures/mock-program/mock_program.so"
    ));
    env.svm
        .add_program(mock_client::mock_program::ID, bytes)
        .unwrap();
}

fn initialize_mock_token_pool(
    env: &mut Env,
    mock_client: &mock_client::MockClient,
    token_owner: Pubkey,
) -> TokenPoolFixture {
    let deposit_mint = Keypair::new();
    let mint_len = spl_token::state::Mint::LEN;
    let create_mint = system_instruction::create_account(
        &env.admin.pubkey(),
        &deposit_mint.pubkey(),
        env.svm.minimum_balance_for_rent_exemption(mint_len),
        mint_len as u64,
        &token::ID,
    );
    send_tx_expect_ok(&mut env.svm, create_mint, &[&env.admin, &deposit_mint]);

    let initialize_mint = spl_token::instruction::initialize_mint2(
        &token::ID,
        &deposit_mint.pubkey(),
        &env.admin.pubkey(),
        None,
        TOKEN_DECIMALS,
    )
    .unwrap();
    send_tx_expect_ok(&mut env.svm, initialize_mint, &[&env.admin]);

    let user_deposit_account =
        mock_client::user_deposit_account(&token_owner, &deposit_mint.pubkey());
    let create_user_deposit_account = create_associated_token_account(
        &env.admin.pubkey(),
        &token_owner,
        &deposit_mint.pubkey(),
        &token::ID,
    );
    send_tx_expect_ok(&mut env.svm, create_user_deposit_account, &[&env.admin]);

    let mint_deposit_tokens = spl_token::instruction::mint_to(
        &token::ID,
        &deposit_mint.pubkey(),
        &user_deposit_account,
        &env.admin.pubkey(),
        &[],
        INITIAL_TOKEN_BALANCE,
    )
    .unwrap();
    send_tx_expect_ok(&mut env.svm, mint_deposit_tokens, &[&env.admin]);

    let initialize_pool = mock_client.initialize_pool(env.admin.pubkey(), deposit_mint.pubkey());
    send_tx_expect_ok(&mut env.svm, initialize_pool, &[&env.admin]);

    let pool = mock_client::pool_pda(&deposit_mint.pubkey());
    let lp_mint = mock_client::lp_mint_pda(&pool);
    let vault = mock_client::vault_address(&pool, &deposit_mint.pubkey());
    let user_lp_account = mock_client::user_lp_account(&token_owner, &lp_mint);
    let create_user_lp_account = create_associated_token_account(
        &env.admin.pubkey(),
        &token_owner,
        &lp_mint,
        &token_2022::ID,
    );
    send_tx_expect_ok(&mut env.svm, create_user_lp_account, &[&env.admin]);

    TokenPoolFixture {
        deposit_mint: deposit_mint.pubkey(),
        lp_mint,
        vault,
        user_deposit_account,
        user_lp_account,
    }
}

fn create_approved_session(
    env: &mut Env,
    client: &client::R3SessionKeysClient,
    allowed_discriminators: Vec<u8>,
    discriminator_size: u8,
) -> (Keypair, Pubkey) {
    let smart_wallet_owner = Keypair::new();
    let session_key = Keypair::new();
    let (create_wallet, user_smart_wallet, _) =
        client.create_smart_wallet(env.admin.pubkey(), smart_wallet_owner.pubkey());
    send_tx_expect_ok(&mut env.svm, create_wallet, &[&env.admin]);

    let expires_at = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs() as i64
        + 1000;
    let (create_session, _, _) = client.create_session(
        env.admin.pubkey(),
        user_smart_wallet,
        session_key.pubkey(),
        mock_client::mock_program::ID,
        expires_at,
        allowed_discriminators,
        discriminator_size,
    );
    send_tx_expect_ok(&mut env.svm, create_session, &[&env.admin]);

    let approve_session = client.approve_session(
        smart_wallet_owner.pubkey(),
        user_smart_wallet,
        session_key.pubkey(),
    );
    send_tx_expect_ok(
        &mut env.svm,
        approve_session,
        &[&env.admin, &smart_wallet_owner],
    );

    (session_key, user_smart_wallet)
}

fn token_account(env: &Env, address: &Pubkey) -> TokenAccount {
    let account = env.svm.get_account(address).unwrap();
    let mut data: &[u8] = &account.data;
    TokenAccount::try_deserialize(&mut data).unwrap()
}

fn mint_supply(env: &Env, address: &Pubkey) -> u64 {
    let account = env.svm.get_account(address).unwrap();
    let mut data: &[u8] = &account.data;
    Mint::try_deserialize(&mut data).unwrap().supply
}

#[test]
fn test_execute_mock_deposit_with_smart_wallet_tokens() {
    let client = client::R3SessionKeysClient::new();
    let mut env = Env::new();
    let initialize_ix = client.initialize(env.admin.pubkey());
    send_tx_expect_ok(&mut env.svm, initialize_ix, &[&env.admin]);

    load_mock_program(&mut env);
    let mock_client = mock_client::MockClient::new();
    let deposit_discriminator = mock_client::deposit_discriminator().to_vec();
    let (session_key, user_smart_wallet) = create_approved_session(
        &mut env,
        &client,
        deposit_discriminator.clone(),
        u8::try_from(deposit_discriminator.len()).unwrap(),
    );

    let session_executor = env.admin.pubkey();
    let token_pool_fixture = initialize_mock_token_pool(&mut env, &mock_client, user_smart_wallet);
    let deposit_account_before = token_account(&env, &token_pool_fixture.user_deposit_account);
    assert_eq!(deposit_account_before.owner, user_smart_wallet);
    assert_eq!(deposit_account_before.amount, INITIAL_TOKEN_BALANCE);

    // The smart wallet is the mock-program user. execute_with_session strips its
    // is_signer flag for the outer tx and restores it on the CPI via invoke_signed.
    let deposit_ix =
        mock_client.deposit(user_smart_wallet, token_pool_fixture.deposit_mint, DEPOSIT_AMOUNT);
    assert!(deposit_ix.data.starts_with(&deposit_discriminator));
    let execute = client.execute_with_session(
        session_executor,
        session_key.pubkey(),
        user_smart_wallet,
        deposit_ix,
    );
    send_tx_expect_ok(&mut env.svm, execute, &[&env.admin, &session_key]);

    assert_eq!(
        token_account(&env, &token_pool_fixture.user_deposit_account).amount,
        INITIAL_TOKEN_BALANCE - DEPOSIT_AMOUNT
    );
    assert_eq!(
        token_account(&env, &token_pool_fixture.vault).amount,
        DEPOSIT_AMOUNT
    );
    assert_eq!(
        token_account(&env, &token_pool_fixture.user_lp_account).amount,
        DEPOSIT_AMOUNT
    );
    assert_eq!(mint_supply(&env, &token_pool_fixture.lp_mint), DEPOSIT_AMOUNT);
}

#[test]
fn test_execute_mock_withdraw_not_allowed_by_deposit_only_session() {
    let client = client::R3SessionKeysClient::new();
    let mut env = Env::new();
    let initialize_ix = client.initialize(env.admin.pubkey());
    send_tx_expect_ok(&mut env.svm, initialize_ix, &[&env.admin]);

    load_mock_program(&mut env);
    let mock_client = mock_client::MockClient::new();
    let deposit_discriminator = mock_client::deposit_discriminator().to_vec();
    let (session_key, user_smart_wallet) = create_approved_session(
        &mut env,
        &client,
        deposit_discriminator.clone(),
        u8::try_from(deposit_discriminator.len()).unwrap(),
    );

    let session_executor = env.admin.pubkey();
    let token_pool_fixture = initialize_mock_token_pool(&mut env, &mock_client, user_smart_wallet);
    let deposit_ix =
        mock_client.deposit(user_smart_wallet, token_pool_fixture.deposit_mint, DEPOSIT_AMOUNT);
    let execute_deposit = client.execute_with_session(
        session_executor,
        session_key.pubkey(),
        user_smart_wallet,
        deposit_ix,
    );
    send_tx_expect_ok(&mut env.svm, execute_deposit, &[&env.admin, &session_key]);

    let balances_before = (
        token_account(&env, &token_pool_fixture.user_deposit_account).amount,
        token_account(&env, &token_pool_fixture.vault).amount,
        token_account(&env, &token_pool_fixture.user_lp_account).amount,
        mint_supply(&env, &token_pool_fixture.lp_mint),
    );

    let withdraw_discriminator = mock_client::withdraw_discriminator();
    let withdraw_ix =
        mock_client.withdraw(user_smart_wallet, token_pool_fixture.deposit_mint, DEPOSIT_AMOUNT);
    assert!(withdraw_ix.data.starts_with(withdraw_discriminator));
    assert!(!withdraw_ix.data.starts_with(&deposit_discriminator));
    let execute_withdraw = client.execute_with_session(
        session_executor,
        session_key.pubkey(),
        user_smart_wallet,
        withdraw_ix,
    );
    let error = send_tx_expect_error(&mut env.svm, execute_withdraw, &[&env.admin, &session_key]);

    assert!(
        error.contains("Error Code: NotAllowedInstructionDiscriminator"),
        "{error}"
    );
    let balances_after = (
        token_account(&env, &token_pool_fixture.user_deposit_account).amount,
        token_account(&env, &token_pool_fixture.vault).amount,
        token_account(&env, &token_pool_fixture.user_lp_account).amount,
        mint_supply(&env, &token_pool_fixture.lp_mint),
    );
    assert_eq!(balances_after, balances_before);
}
